mypyc/irbuild/specialize.py

```diff
@@ -135,13 +141,47 @@ def translate_safe_generator_call(
                     + [builder.accept(arg) for arg in expr.args[1:]]),
                 builder.node_type(expr), expr.line, expr.arg_kinds, expr.arg_names)
         else:
+            if callee.fullname == "builtins.tuple":
+                val = tuple_from_generator_helper(builder, expr.args[0])
+                if val is not None:
+                    return val
             return builder.call_refexpr_with_args(
                 expr, callee,
                 ([translate_list_comprehension(builder, expr.args[0])]
                     + [builder.accept(arg) for arg in expr.args[1:]]))
     return None
 
 
+def tuple_from_generator_helper(builder: IRBuilder,
+                                gen: GeneratorExpr) -> Optional[Value]:
+
+    if len(gen.sequences) == 1 and len(gen.condlists[0]) == 0:
+        # Currently we only optimize for simplest generator expression
+        rtype = builder.node_type(gen.sequences[0])
+        if ((is_list_rprimitive(rtype) or is_tuple_rprimitive(rtype))
+                and is_int_rprimitive(builder.node_type(gen.indices[0]))):
+            tuple_ops = builder.builder.new_tuple_with_length(builder.accept(gen.sequences[0]),
+                                                              gen.line)
+            index, expr = gen.indices[0], gen.sequences[0]
+
+            def set_tuple_item() -> None:
+                e = builder.accept(gen.left_expr)
+                index_val = builder.accept(index)
```

Comment:
I don't think that this works as expected. The naming is pretty confusing, but I think that this is actually a sequence item (`x` in `for x in my_list`), not the index. Actually `for_loop_helper` doesn't seem to give you access to the sequence index. That probably means that `for_loop_helper` isn't a good match for this use case, unfortunately. 

Directly using `ForEnumerate` or maybe `ForRange` could be better options. If using `ForEnumerate`, you'll have to create a new lvalue for `index1` (a register). If using `ForRange`, you'll need to use something similar to what `transform_assignment_stmt` does to assign the sequence item to the lvalue(s). The key part is this:

```
    target = builder.get_assignment_target(lvalue)
    builder.assign(target, rvalue_reg, line)
```

Reply:
I have checked the generated IR, it seems that it dose work.

Reply:
As discussed above, can you check that this works for types such as `List[str]`?

Reply:
Well, I didn't consider that. > <
I would add a new helper function similar to `for_loop_helper`.

Reply:
Yeah, that should work. One option would be to have a similar helper that gives you access both to the index (starting from 0) and the item value. You don't need to support everything that `for_loop_helper` supports since this helper would be used in more specialized cases.